app/src/main/proto/vss.proto

```diff
@@ -293,7 +293,40 @@ message KeyValue {
   int64 version = 2;
 
   // Object value in bytes which is stored (in put) and fetched (in get).
-  // Clients must encrypt this blob client-side before sending it over the wire to server in order
-  // to preserve privacy and security.
+  // Clients must encrypt secret contents of this blob client-side before sending it over the wire
+  // to server in order to preserve privacy and security.
+  // Clients may use a `Storable` object on client-side and set it here.
   bytes value = 3;
 }
+
+// Represents a storable object that can be serialized and stored as value in `PutObjectRequest`.
+// Server does not use or read anything from Storable, clients may use fields as required.
+// Only provided as a helper object for ease of use by clients.
+message Storable {
```

Comment:
Sure, but it's a data abstraction representing encrypted data. The fact that it also contains metadata doesn't change that. The name `Storable` can describe anything and could have just as easily been named `Data`.

Reply:
> Storable is something that will be stored, whereas data is very non-specific. (it is a naming convention for objects that are to be stored in database, [link](https://docs.oracle.com/en/industries/communications/billing-revenue/12.0/dev-guide/understanding-storable-classes1.html#GUID-A165AD0B-E419-4ACA-936C-DB053BED819B))
> 

What would you name a message that contained non-encyrpted data to store? This message precludes using it for that but is named otherwise.

The example you cite is used to map class fields to database columns. This doesn't have a similar usage.

> * I refrain from naming it encryptedData because it is not encrypted. (could be confusing)

I'm confused now... Doesn't it contain encrypted data and related metadata?

> Also, encryption is not the only characteristic of inner_data just one of them, for example if we add compression. We don't want to add EncryptedCompressedData.

If you want to call it `Storable`, I feel that the value field should have a generic name and any other optional fields for metadata should be separate messages with their own set of fields (e.g., an `encryption` field holding an `EncryptionMetadata` message). Then a reader would examine the metadata fields in order to perform operations (e.g., decrypt, uncompress) on the value.

Reply:
> I don't understand this recommendation fully but feel this is overly complex.
> 
> Will try to discuss with you offline.

If you use optional message fields, you can support cases where the data is encrypted but is not compressed, for instance. Then the client library can use existence checks to decrypt, uncompress, etc.

Reply:
Thanks, but looks like this wasn't addressed. The idea is to have one optional message field using a new message as the type for the encryption metadata fields. The client code for encryption would return this to set on the `Storable`.

Then if compression is added later, another message for it's metadata would be added using a separate field. And the client code for compression would likewise return its own metadata message to have set on `Storable`.

Reply:
Yeah, I'd prefer a separate message as it can define everything that must be set. With separate fields, you can have a weird state where only some of the fields are set even if all of them must be. Or if none of them should be set, you'd have to give them default values, in Rust, I'd imagine? How do optional fields work in the Rust API, come to think of it? Do they use `Option`?

It can also be initialized inline when constructing the `Storable`, so I don't think it's much boilerplate.

```rust
Storable {
	value: cipher_blob,
	encryption_metadata: EncryptionMetadata {
		nonce: nonce.to_vec(),
		tag: tag.to_vec(),
		cipher_format: CHACHA20_CIPHER_NAME.to_string(),
	},
}
```